src/mlpack/core/tree/greedy_single_tree_traverser_impl.hpp

```diff
@@ -24,26 +24,48 @@ template<typename TreeType, typename RuleType>
 GreedySingleTreeTraverser<TreeType, RuleType>::GreedySingleTreeTraverser(
     RuleType& rule) :
     rule(rule),
-    numPrunes(0)
+    numPrunes(0),
+    k(0)
 { /* Nothing to do. */ }
 
 template<typename TreeType, typename RuleType>
 void GreedySingleTreeTraverser<TreeType, RuleType>::Traverse(
     const size_t queryIndex,
     TreeType& referenceNode)
 {
-  // Run the base case as necessary for all the points in the reference node.
-  for (size_t i = 0; i < referenceNode.NumPoints(); ++i)
-    rule.BaseCase(queryIndex, referenceNode.Point(i));
+  if(referenceNode.IsLeaf())
+  {
+    // Run the base case as necessary for all the points in the reference node.
+    for (size_t i = 0; i < referenceNode.NumPoints(); ++i)
+      rule.BaseCase(queryIndex, referenceNode.Point(i));
+    return;
+  }
+
+  size_t bestChild = rule.GetBestChild(queryIndex, referenceNode);
+  size_t numDescendants;
 
-  if (!referenceNode.IsLeaf())
+  // Check that referencenode is not a leaf node while calculating number of
+  // descendants of it's best child.
+  if(!referenceNode.IsLeaf())
+    numDescendants = referenceNode.Child(bestChild).NumDescendants();
+  else
+    numDescendants = referenceNode.NumPoints();
+
+  // If number of descendants are more than k than we can go along with
+  // best child otherwise we need to traverse for each descendant to
+  // ensure that we get at least k nearest neighbors..
+  if (numDescendants > k)
   {
     // We are prunning all but one child.
     numPrunes += referenceNode.NumChildren() - 1;
     // Recurse the best child.
-    size_t bestChild = rule.GetBestChild(queryIndex, referenceNode);
     Traverse(queryIndex, referenceNode.Child(bestChild));
   }
+  else
+  {
+    for (size_t i = 0; i < referenceNode.NumChildren(); ++i)
+      Traverse(queryIndex, referenceNode.Child(i));
+  }
```

Comment:
It looks like this code will traverse even when the node is a leaf.  Instead if we do the pruning, we can just loop over all the points directly.  I think you could structure more like this:

```
if (!referenceNode.IsLeaf())
{
  const size_t bestChild = ... ;
  if (referenceNode.Child(bestChild).NumDescendants() > (k - baseCases))
  {
    Traverse(queryIndex, referenceNode.Child(bestChild);
  }
  else
  {
    for (size_t i = 0; i < referenceNode.NumDescendants(); ++i) // I guess we could terminate at k points also
    {
      rule.BaseCase(queryIndex, referenceNode.Descendant(i));
    }
  }
}
```

What do you think?  The idea there might not be quite right but I hope it is enough to go on.

Reply:
Curently this is not happening, but will take place once `referenceNode.IsLeaf()` condition is removed while calculating base_cases. So, this will be a good measure to impose `!referenceNode.IsLeaf()` condition here. But I don't quite understand the use of `(k - baseCases) `. I think it will cause a bit of uncertainity here. It think it will work fine to just use `referenceNode.Child(bestChild).NumDescendants() > k` as in case we do not determine k the code will just convert back to previous one.  And tracking BaseCases here may not be optimum, as we may have to declare baseCases globally or will have to change the class definition. Let, me know if I interpreted it correctly.
```

for (size_t i = 0; i < referenceNode.NumDescendants(); ++i) // I guess we could terminate at k points also
    {
      rule.BaseCase(queryIndex, referenceNode.Descendant(i));
    }
```
Nice, using NumDescendant is better way of doing the same and now we can just break the loop at ``i==k``. But breaking at` i==k` may give somewhat worse results for nearest neighbours. But as we know this is greedy technique, so everything is fair in greedy till we pick the best option before us  :)